### PR TITLE
PYTHON-3047 Clean up handling of pyarrow library

### PR DESCRIPTION
**Pull request by blink1073**
- [x] Modernize to use PEP517-compliant build system
- [x] Copy the arrow libs independently in wheel (but allow it to be overridden for conda-forge so we can use `arrow-cpp`)
- [x] Link to the arrow library files directly instead of using `pyarrow.create_library_symlinks()`
- [x] Updated docs
- [x] Tested locally using a conda [recipe](https://gist.github.com/blink1073/b56e7f1389a2b54fd357d96e544398f7)

### CODE REVIEW

**Comment by blink1073** [comment]
Converting back to draft until I have a working conda recipe

File: bindings/python/setup.py [hunk 1]
```diff
@@ -93,6 +104,10 @@ def append_arrow_flags(module):
 
 
 def get_extension_modules():
+    try:
+        from Cython.Build import cythonize
+    except ImportError:
+        return []
```
Comment (ShaneHarvey) [inline bindings/python/setup.py:110]: Why this change?

File: bindings/python/MANIFEST.in [hunk 2]
```diff
@@ -2,13 +2,19 @@ include README.rst
 include LICENSE
 include build-libbson.sh
 include release.sh
```
Comment (ShaneHarvey) [inline bindings/python/MANIFEST.in:4]: Should we remove the release script here?

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: bindings/python/pymongoarrow/__init__.py [hunk 3]
```diff
@@ -15,9 +15,10 @@
 # We must import pyarrow before attempting to load the Cython module.
 import pyarrow
 
-from pymongoarrow.lib import libbson_version
 from pymongoarrow.version import __version__, _MIN_LIBBSON_VERSION
-
+# this line must come second so setuptools can parse the __version__
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/pymongoarrow/__init__.py:19]: It is using an ast parser

File: bindings/python/pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+requires = [
+    "setuptools>=47.9",
+    "wheel",
+    "cython>=0.29",
+    # Must be kept in sync with the `install_requires` in `setup.cfg`
+    "pyarrow>=6.0.0,<6.1.0",
+]
+
+[tool.check-manifest]
+ignore = ["pymongoarrow/*.so*.*", "pymongoarrow/*.so*.*", "pymongoarrow/*.dylib"]
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/pyproject.toml:11]: done

File: bindings/python/pymongoarrow/__init__.py [hunk 5]
```diff
@@ -15,9 +15,10 @@
...
+# above without having a built application
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/pymongoarrow/__init__.py:20]: Done

File: bindings/python/setup.py [hunk 6]
```diff
@@ -72,18 +58,43 @@ def append_libbson_flags(module):
...
+    import pyarrow as pa
+
     module.include_dirs.append(np.get_include())
     module.include_dirs.append(pa.get_include())
-    module.libraries.extend(pa.get_libraries())
     module.library_dirs.extend(pa.get_library_dirs())
 
+    # Handle the arrow library files manually
+    # Alternative to using pyarrow.create_library_symlinks()
+    arrow_lib = os.environ.get('MONGO_ARROW_LIBDIR')
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/setup.py:77]: Added docs

File: bindings/python/setup.py [hunk 7]
```diff
@@ -72,18 +58,43 @@ def append_libbson_flags(module):
...
+        arrow_lib = pa.get_library_dirs()[0]
+        if platform == "darwin":
+            pattern = '*.dylib'
+        elif platform == 'linux':
+            pattern = '*.so.*'
+
+    # Handle linking and optionally copy the library files
+    for name in ['libarrow', 'libarrow_python']:
+        path = glob.glob(os.path.join(arrow_lib, name + pattern))[0]
+        if not os.environ.get("MONGO_NO_COPY_ARROW_LIB", False):
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/setup.py:93]: Added docs

File: bindings/python/setup.py (hunk 1 above)
Comment (blink1073) [reply bindings/python/setup.py:110]: Added docs

File: bindings/python/MANIFEST.in (hunk 2 above)
Comment (blink1073) [reply bindings/python/MANIFEST.in:4]: Sounds good

File: .github/workflows/test-python.yml [hunk 8]
```diff
@@ -27,21 +31,17 @@ jobs:
...
+          # Install the library
+          LIBBSON_INSTALL_DIR=$(pwd)/libbson python -m pip install -vvv -e ".[test]"
       - name: Run tests
         run: |
           LD_LIBRARY_PATH=$(pwd)/libbson/lib python -m unittest discover test -v
-        working-directory: ./bindings/python
+      - name: Check the manifest
+        run: |
+          pip install check-manifest
+          check-manifest -v
```
Comment (ShaneHarvey) [inline .github/workflows/test-python.yml:47]: What is check-manifest and why is it needed?

File: bindings/python/setup.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply bindings/python/setup.py:100]: Thx. When we copy the libs, do we need to also update the extra_link_args path to point to the copied versions?

File: bindings/python/setup.py (hunk 1 above)
Comment (ShaneHarvey) [reply bindings/python/setup.py:119]: Ack.

File: bindings/python/setup.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply bindings/python/setup.py:82]: Ack.

**Review by ShaneHarvey — COMMENTED** [review]
2 more questions.

File: bindings/python/pymongoarrow/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/pymongoarrow/__init__.py]: It seems like it is able to parse a portion of the file at a time, when they were in the reverse order it did not work

File: bindings/python/setup.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/setup.py:100]: Good call, updated

File: .github/workflows/test-python.yml (hunk 8 above)
Comment (blink1073) [reply .github/workflows/test-python.yml:47]: It makes sure that you are intentionally handling all of the files in version control and as built so you don't accidentally add/subtract files.

Comment (ShaneHarvey) [reply .github/workflows/test-python.yml:47]: Ack.

File: bindings/python/MANIFEST.in [hunk 9]
```diff
@@ -1,14 +1,21 @@
...
+
+exclude THIRD-PARTY-NOTICES
+exclude release.sh
+exclude addtags.py
+exclude benchmark.py
 
 graft pymongoarrow
 
-prune test
+recursive-exclude test *
```
Comment (ShaneHarvey) [inline bindings/python/MANIFEST.in:13]: It appears from the logs that test test/ dir now gets included in the wheel build. Is somethign off here?
```
copying build/lib.linux-x86_64-3.10/pymongoarrow/bson.pyi -> build/bdist.linux-x86_64/wheel/pymongoarrow
creating build/bdist.linux-x86_64/wheel/test
copying build/lib.linux-x86_64-3.10/test/test_schema.py -> build/bdist.linux-x86_64/wheel/test
copying build/lib.linux-x86_64-3.10/test/test_arrow.py -> build/bdist.linux-x86_64/wheel/test
copying build/lib.linux-x86_64-3.10/test/test_pymongoarrow.py -> build/bdist.linux-x86_64/wheel/test
copying build/lib.linux-x86_64-3.10/test/__init__.py -> build/bdist.linux-x86_64/wheel/test
copying build/lib.linux-x86_64-3.10/test/test_datetime.py -> build/bdist.linux-x86_64/wheel/test
copying build/lib.linux-x86_64-3.10/test/test_monkey.py -> build/bdist.linux-x86_64/wheel/test
copying build/lib.linux-x86_64-3.10/test/utils.py -> build/bdist.linux-x86_64/wheel/test
copying build/lib.linux-x86_64-3.10/test/test_builders.py -> build/bdist.linux-x86_64/wheel/test
copying build/lib.linux-x86_64-3.10/test/test_pandas.py -> build/bdist.linux-x86_64/wheel/test
copying build/lib.linux-x86_64-3.10/test/test_numpy.py -> build/bdist.linux-x86_64/wheel/test
copying build/lib.linux-x86_64-3.10/test/test_libbson.py -> build/bdist.linux-x86_64/wheel/test
copying build/lib.linux-x86_64-3.10/test/test_bson.py -> build/bdist.linux-x86_64/wheel/test
...
```
https://github.com/mongodb-labs/mongo-arrow/runs/4527432236?check_suite_focus=true#step:3:607

Comment (blink1073) [reply bindings/python/MANIFEST.in:13]: Fixed in latest push

**Comment by blink1073** [comment]
I also cleaned up the library file handling in the latest commit

File: bindings/python/setup.py [hunk 10]
```diff
@@ -72,18 +58,48 @@ def append_libbson_flags(module):
...
+    # Alternative to using pyarrow.create_library_symlinks().
+    # You can use MONGO_ARROW_LIBDIR to explicitly set the location of the 
+    # arrow libraries (for instance in a conda build).
+    # We first check for an unmodified path to the library,
+    # then look for a library file with a version modifier, e.g. libarrow.600.dylib.
+    arrow_lib = os.environ.get('MONGO_ARROW_LIBDIR', pa.get_library_dirs()[0])
+    if platform == "darwin":
+        patterns = ['.dylib', '.[0-9][0-9][0-9].dylib']
+    elif platform == 'linux':
+        patterns = ['.so', '.so.[0-9][0-9][0-9]']
```
Comment (ShaneHarvey) [inline bindings/python/setup.py:85]: What happens when pyarrow 10 comes out? would this need to change to `'.so.[0-9][0-9][0-9][0-9]'`? Is there a way to future proof that?

Comment (blink1073) [reply bindings/python/setup.py]: Updated to add regex handling

Comment (blink1073) [reply bindings/python/setup.py]: Updated again to working glob syntax with wildcards

**Review by ShaneHarvey — APPROVED** [review]
LGTM!
